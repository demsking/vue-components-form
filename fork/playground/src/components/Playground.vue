<template>
  <div class="playground">
    <div class="playground__input">
      <div class="playground__input__schema">
        <h2>Schema</h2>
        <div class="playground__input__schema__editor">
          <PrismEditor
            :code="rawSchema"
            language="json"
            @change="setRawSchema"/>
        </div>
      </div>
      <div class="playground__input__rendering">
        <h2>Rendering</h2>
        <div ref="form" class="playground__input__rendering__viewport">
          <FormSchema :schema="schema" :descriptor="descriptor" v-model="model" @ready="generateCode" @submit.prevent>
            <div class="buttons">
              <button type="submit">Subscribe</button>
            </div>
          </FormSchema>
        </div>
      </div>
      <div class="playground__input__model">
        <h2>Model</h2>
        <pre>{{ model }}</pre>
      </div>
    </div>
    <div class="playground__output">
      <h2>Generated HTML</h2>
      <div class="playground__output__code">
        <PrismEditor :code="code" language="html" readonly/>
      </div>
    </div>
  </div>
</template>

<script>
  import '../libs/Prism';
  import '../styles/Prism.css';

  import { html } from 'js-beautify';

  import PrismEditor from 'vue-prism-editor';
  import FormSchema from '../../../dist/FormSchema.esm';
  import Schema from '../schema/newsletter';

  export default {
    components: { PrismEditor, FormSchema },
    data: () => ({
      model: undefined,
      rawSchema: JSON.stringify(Schema, null, 2),
      code: '',
      formatter: null,
      descriptor: {
        properties: {
          name: {
            properties: {
              first_name: {
                attrs: {
                  placeholder: 'Your First Name',
                  title: 'Please enter your first name'
                }
              },
              middle_name: {
                attrs: {
                  placeholder: 'Your Middle Name',
                  title: 'Please enter your middle name'
                }
              },
              last_name: {
                attrs: {
                  placeholder: 'Your Last Name',
                  title: 'Please enter your last name'
                }
              }
            }
          },
          email: {
            attrs: {
              placeholder: 'Your Email',
              title: 'Please enter your email'
            }
          },
          day: {
            label: 'Newsletter Day',
            attrs: {
              placeholder: 'Select your list subscription',
              title: 'Please select your list subscription'
            },
            labels: {
              monday: 'Monday',
              tuesday: 'Tuesday',
              wednesday: 'Wednesday',
              thursday: 'Thursday',
              friday: 'Friday',
              saturday: 'Saturday',
              sunday: 'Sunday'
            }
          },
          source: {
            kind: 'textarea',
            label: 'Source',
            description: 'Ex. Using the NPM Search Engine',
            attrs: {
              placeholder: 'How did you hear about us?'
            }
          },
          password: {
            attrs: {
              type: 'password'
            }
          },
          frequence: {
            labels: {
              daily: 'Daily',
              weekly: 'Weekly'
            }
          }
        }
      }
    }),
    computed: {
      schema() {
        return JSON.parse(this.rawSchema);
      }
    },
    methods: {
      setRawSchema(value) {
        try {
          JSON.parse(value);

          this.rawSchema = value;
        } catch(e) {}
      },
      generateCode() {
        const code = this.$refs.form.innerHTML;

        this.code = html(code.replace(/\s*<!---->/g, ''));
      }
    }
  }
</script>

<style>
  .playground {
    text-align: left;
    margin: auto;
    /* display: flex; */
    flex-direction: column;
    color: #f5f5f5;
    background-color: #333333;
    overflow: hidden;

    position: relative;
    height: 100%;
  }

  .playground h2 {
    margin: 0;
    padding: 10px;
    font-size: .90em;
    font-weight: 400;
    width: 100%;
  }

  .playground pre {
    margin: 0;
  }

  .playground__input {
    display: flex;
    flex-direction: row;
    /* max-height: 550px; */

    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 60%;
  }

  .playground h2,
  .playground__input {
    border-bottom: 1px solid rgba(255, 255, 255, .1);
  }

  .playground__input__schema {
    display: flex;
    flex-direction: column;
  }

  .playground__input__schema__editor {
    flex: 1;
    display: flex;
    align-items: flex-start;
    max-width: 500px;
    overflow: auto;
  }

  .playground__input__rendering {
    border-left: 1px solid rgba(255, 255, 255, .1);
    border-right: 1px solid rgba(255, 255, 255, .1);
  }

  .playground__input__rendering,
  .playground__input__model {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    overflow: hidden;
  }

  .playground__input__rendering__viewport {
    padding: 20px;
    overflow: auto;
    margin: auto;
    color: #333;
    background-color: #f5f5f5;
  }

  .playground__input__rendering legend {
    font-size: 1.7em;
    text-align: center;
    margin-top: 0;
    margin-bottom: .2em
  }

  .playground__input__rendering legend + p {
    display: block;
    text-align: center;
    margin-bottom: 1.2em
  }

  .playground__input__rendering input,
  .playground__input__rendering textarea,
  .playground__input__rendering select {
    display: block;
  }

  .playground__input__model {
    margin: 0;
    min-width: 100px;
  }

  .playground__output {
    position: absolute;
    top: 60%;
    left: 0;
    width: 100%;
    height: 40%;

    overflow: hidden;
    display: flex;
    flex-direction: column;
  }

  .playground__output__code {
    flex: 1;
    display: flex;
    overflow: auto;
  }

  [data-fs-field] {
    display: flex;
    flex-direction: row;
    margin-bottom: 5px;
  }

  [data-fs-field] label {
    display: block;
    width: 120px;
    text-align: right;
    margin-right: 10px
  }

  [data-fs-kind="enum"] [data-fs-field] {
    flex-direction: row-reverse;
  }

  [data-fs-kind="enum"] [data-fs-field] label {
    text-align: left;
    flex: 1;
  }
</style>
