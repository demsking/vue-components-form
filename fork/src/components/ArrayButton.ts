import { FunctionalComponentOptions } from 'vue';

export const ArrayButton: FunctionalComponentOptions = {
  functional: true,
  name: 'ArrayButton',
  render(h, { data, listeners }) {
    return h('button', {
      attrs: { type: 'button', ...data.attrs },
      on: listeners
    }, 'Add');
  }
};
