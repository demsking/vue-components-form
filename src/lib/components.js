import { INPUT_TYPES } from './parser'

const BLOCK_TYPES = [INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT]

const Input = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const element = BLOCK_TYPES.includes(field.attrs.type)
      ? h(field.attrs.type, data, slots().default)
      : h('input', data)
    const nodes = [ element ]

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label || (field.isArrayField && !BLOCK_TYPES.includes(field.attrs.type))) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const CheckboxGroup = {
  functional: true,
  render (h, { data, slots }) {
    const name = data.field.attrs.name
    const description = data.field.description

    const children = [ ...slots().default ]

    if (description) {
      children.unshift(h('legend', description))

      delete data.field.description
    }

    return h('fieldset', { attrs: { name } }, children)
  }
}

const ArrayInputs = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-array-inputs': true
      }
    }, slots().default)
  }
}

const Label = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const attrs = {
      'data-fs-field': field.attrs.id,
      'data-fs-required': field.attrs.required
    }

    return h('div', { attrs }, [
      h('label', {
        attrs: {
          for: field.attrs.id
        }
      }, field.label),
      h('div', {
        attrs: {
          'data-fs-field-input': field.attrs.id
        }
      }, slots().default)
    ])
  }
}

const Fieldset = {
  functional: true,
  render (h, { data, slots }) {
    const field = data.field
    const newItemButton = data.newItemButton
    const nodes = []

    if (field.isArrayField) {
      if (newItemButton) {
        nodes.push(h(ArrayInputs, slots().default))
        nodes.push(h(ArrayButton, newItemButton))
      } else {
        nodes.push(h(CheckboxGroup, data, slots().default))
      }
    } else {
      nodes.push(h('fieldset', slots().default))
    }

    if (field.description) {
      nodes.push(h('small', field.description))
    }

    if (!field.label) {
      if (nodes.length === 1) {
        return nodes[0]
      }
      return nodes
    }

    return h(Label, data, nodes)
  }
}

const ArrayButton = {
  functional: true,
  render (h, { data }) {
    return h('button', {
      attrs: { type: 'button', ...data.props },
      on: data.on
    }, 'Add')
  }
}

const ErrorElement = {
  functional: true,
  render (h, { slots }) {
    return h('div', {
      attrs: {
        'data-fs-error': true
      }
    }, slots().default)
  }
}

const TAGS = {
  title: 'h1',
  description: 'p',
  formwrapper: 'div',
  form: 'form',
  text: Input,
  option: 'option',
  fieldset: Fieldset,
  error: ErrorElement
}

export class Components {
  constructor () {
    this.$ = {}

    for (let type in TAGS) {
      this.set(type, TAGS[type], true)
    }
  }

  set (type, component, native = false) {
    this.$[type] = { type, native, component }
  }

  input ({ field, fieldParent = null, listeners = {} }) {
    const { type } = field.attrs
    const element = field.hasOwnProperty('items') && groupedArrayTypes.includes(type)
      ? this.$.fieldset
      : this.$[type] || this.$[INPUT_TYPES.TEXT]

    const data = {
      field,
      fieldParent,
      components: this.components,
      attrs: {},
      props: {},
      domProps: {},
      [argName(element)]: { ...field.attrs },
      on: listeners
    }

    return { element, data }
  }
}

export function argName (el) {
  return el.native ? 'attrs' : 'props'
}

export const groupedArrayTypes = [
  INPUT_TYPES.RADIO, INPUT_TYPES.CHECKBOX, 'input', INPUT_TYPES.TEXTAREA
]

export const fieldTypesAsNotArray = [
  INPUT_TYPES.RADIO, INPUT_TYPES.TEXTAREA, INPUT_TYPES.SELECT
]

export const inputName = (field, index) => `${field.attrs.name}-${index}`
