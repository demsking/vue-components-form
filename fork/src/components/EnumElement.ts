import { FunctionalComponentOptions } from 'vue';
import { HelperElement } from "@/components/HelperElement";

export const EnumElement: FunctionalComponentOptions = {
  name: 'EnumElement',
  functional: true,
  render(h, { data, props }) {
    const { components } = props;

    const nodes = props.field.fields.map(({ field }) => {
      const type = field.kind;

      return h(components.get(type), {
        attrs: field.attrs.input,
        props: {
          field: field,
          value: field.model,
          components: props.components
        }
      });
    });

    const helper = h(HelperElement, data);

    if (helper) {
      nodes.unshift(helper);
    }

    if (props.field.label) {
      const legend = h('legend', {
        attrs: props.field.attrs.label
      }, props.field.label);

      nodes.unshift(legend);
    }

    return h('fieldset', {
      attrs: props.field.attrs.input
    }, nodes);
  }
};
