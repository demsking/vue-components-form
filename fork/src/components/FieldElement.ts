import { HelperElement } from "@/components/HelperElement";
import { Dictionary, FieldComponent } from '@/types';

export const FieldElement: FieldComponent = {
  name: 'FieldElement',
  functional: true,
  render(h, { data, props, slots }) {
    const attrs: Dictionary = {
      'data-fs-kind': props.field.kind,
      'data-fs-field': props.field.name,
      'data-fs-required': props.field.required
    };

    const labelElement = h('label', {
      attrs: props.field.attrs.label
    }, props.field.descriptor.label);

    const fieldNodes = [
      ...slots().default,
      h(HelperElement, data)
    ];

    const fieldElement = h('div', {
      attrs: {
        'data-fs-input': (props.field.attrs.input as any).type || props.field.kind
      }
    }, fieldNodes);

    const nodes = [ labelElement, fieldElement ];

    return h('div', { attrs }, nodes);
  }
};
