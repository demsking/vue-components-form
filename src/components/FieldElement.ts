import { HelperElement } from '@/components/HelperElement';
import { Dict, FieldComponent } from '@/types';

export const FieldElement: FieldComponent = {
  name: 'FieldElement',
  functional: true,
  render(h, { data, props, children }) {
    const field = props.field;
    const descriptor = props.descriptor;

    if (field.kind === 'hidden') {
      return children;
    }

    const type = field.attrs.type || field.kind;
    const attrs: Dict = {
      'data-fs-kind': field.kind,
      'data-fs-type': type,
      'data-fs-field': field.name,
      'data-fs-required': field.required,
      'data-fs-horizontal': descriptor.props.horizontal
    };

    const labelElement = h('label', {
      attrs: descriptor.labelAttrs
    }, descriptor.label);

    const fieldElement = h('div', {
      attrs: {
        'data-fs-input': type
      }
    }, children);

    const nodes = [ fieldElement ];
    const helperNode = h(HelperElement, data);

    if (helperNode.tag) {
      nodes.push(helperNode);
    }

    if (field.messages.length) {
      const MessageElement = descriptor.components.get('message');

      field.messages.forEach(({ text, type = 3 }) => {
        nodes.push(h(MessageElement, {
          props: { text, type }
        }, text));
      });
    }

    const wrapperElement = nodes.length === 1
      ? nodes
      : h('div', {
        attrs: {
          'data-fs-wrapper': nodes.length
        }
      }, nodes);

    return h('div', { attrs }, [ labelElement, wrapperElement ]);
  }
};
