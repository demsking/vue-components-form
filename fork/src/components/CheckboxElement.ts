import { CreateStateInput } from '@/lib/NativeElementsLib';

export const CheckboxElement = CreateStateInput('checkbox');
