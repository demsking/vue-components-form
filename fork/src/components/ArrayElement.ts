import { FunctionalComponentOptions } from 'vue';
import { FieldElement } from './FieldElement';
import { ArrayButton } from './ArrayButton';

function renderField(h, field, props) {
  return h(field.component, {
    attrs: field.attrs.input,
    props: {
      ...props,
      field: field,
      value: field.model
    }
  });
}

export const ArrayElement: FunctionalComponentOptions = {
  name: 'ArrayElement',
  functional: true,
  render(h, { props }) {
    const attrs = props.field.attrs.input;
    const nodesx = props.field.items.map((field) => {
      return h(field.component, {
        attrs: field.attrs.input,
        props: {
          ...props,
          field: field,
          value: field.model
        }
      });
    });

    const max = props.field.maxItems
      ? props.field.maxItems + props.field.additionalItems.length
      : -1;

    const limit = max === -1
      ? props.field.count
      : props.field.count < props.field.maxItems
        ? props.field.count
        : props.field.maxItems;

    const nodes = Array(...Array(limit)).map((x, i) => {
      const field = props.field.items[0];

      return renderField(h, field, props);
    });

    if (limit < props.field.count) {
      props.field.additionalItems.forEach((field) =>  {
        nodes.push(renderField(h, field, props));
      });
    }

    const addButtonElement = h(ArrayButton, {
      attrs: {
        disabled: props.field.count === max
      },
      props: props,
      on: {
        click() {
          props.field.count++;
        }
      }
    });

    nodes.push(addButtonElement);

    return h(FieldElement, { attrs, props }, nodes);
  }
};
