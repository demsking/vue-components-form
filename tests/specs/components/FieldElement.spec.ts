import { mount } from '@vue/test-utils';
import { FieldElement } from '@/components/FieldElement';
import { Options } from '../../lib/Options';

const { context, field } = Options.get({
  schema: {
    type: 'string',
    pattern: 'arya|jon',
    minLength: 5,
    maxLength: 15,
    title: 'Name',
    description: 'Your full name'
  },
  model: 'Goku',
  id: 'id-name',
  name: 'name'
});

const stubs: any = {
  HelperElement: true
};

describe('components/FieldElement', () => {
  it('should successfully render component', () => {
    const wrapper = mount(FieldElement, { context, stubs });
    const expected = '<div data-fs-kind="string" data-fs-type="text" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label><div data-fs-wrapper="2"><div data-fs-input="text"></div><p id="id-name-helper" data-fs-helper="true">Your full name</p></div></div>';

    expect(wrapper.is(FieldElement)).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot(expected);
  });

  it('should successfully render component with missing field.attrs.type', () => {
    delete field.attrs.type;

    const wrapper = mount(FieldElement, { context, stubs });
    const expected = '<div data-fs-kind="string" data-fs-type="string" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label><div data-fs-wrapper="2"><div data-fs-input="string"></div><p id="id-name-helper" data-fs-helper="true">Your full name</p></div></div>';

    expect(wrapper.is(FieldElement)).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot(expected);
  });
});
