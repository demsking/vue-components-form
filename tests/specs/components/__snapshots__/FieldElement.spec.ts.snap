// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FieldElement should successfully render component with missing field.attrs.type: <div data-fs-kind="string" data-fs-type="string" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label><div data-fs-wrapper="2"><div data-fs-input="string"></div><p id="id-name-helper" data-fs-helper="true">Your full name</p></div></div> 1`] = `
<div data-fs-kind="string" data-fs-type="string" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label>
  <div data-fs-wrapper="2">
    <div data-fs-input="string"></div>
    <p id="id-name-helper" data-fs-helper="true">Your full name</p>
  </div>
</div>
`;

exports[`components/FieldElement should successfully render component: <div data-fs-kind="string" data-fs-type="text" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label><div data-fs-wrapper="2"><div data-fs-input="text"></div><p id="id-name-helper" data-fs-helper="true">Your full name</p></div></div> 1`] = `
<div data-fs-kind="string" data-fs-type="text" data-fs-field="name"><label id="id-name-label" for="id-name">Name</label>
  <div data-fs-wrapper="2">
    <div data-fs-input="text"></div>
    <p id="id-name-helper" data-fs-helper="true">Your full name</p>
  </div>
</div>
`;
